Sources/Hummingbird/Router/Trie/Trie+serialize.swift

```diff
@@ -14,72 +14,88 @@
 
 import NIOCore
 
-extension BinaryTrie {
+extension RouterTrie {
+    @usableFromInline
     static func serialize(
         _ node: RouterPathTrieBuilder<Value>.Node,
-        trie: inout ByteBuffer,
+        trie: inout Trie,
         values: inout [Value?]
     ) {
-        let binaryTrieNodeIndex = trie.writerIndex
-        trie.reserveBinaryTrieNode()
         // Index where `value` is located
-        let index = UInt16(values.count)
+        let valueIndex = values.count
         values.append(node.value)
 
-        let token: BinaryTrieTokenKind
+        let token: TrieToken
+
+        func setConstant(_ constant: Substring) -> UInt16 {
+            if let index = trie.constants.firstIndex(of: constant) {
+                return UInt16(index)
+            } else {
+                let index = trie.constants.count
+                trie.constants.append(constant)
+                return UInt16(index)
+            }
+        }
+
+        func setParameter(_ parameter: Substring) -> UInt16 {
+            if let index = trie.parameters.firstIndex(of: parameter) {
```

Comment:
You could copy the parameter and constant strings into one big string. You are already using Substrings to store them (to ensure they are next to each other in memory)

Reply:
That's clever! That would allow the CPU caches to work a bit more.

Reply:
```
RouterTrie:LongPaths
╒═══════════════════════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╕
│ Metric                    │      p0 │     p25 │     p50 │     p75 │     p90 │     p99 │    p100 │ Samples │
╞═══════════════════════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╡
│ Malloc (total) *          │      14 │      14 │      14 │      14 │      14 │      14 │      14 │     384 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Throughput (# / s) (K)    │     393 │     392 │     391 │     388 │     384 │     372 │     362 │     384 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (total CPU) (ns) *   │    2564 │    2572 │    2578 │    2591 │    2623 │    2712 │    2773 │     384 │
╘═══════════════════════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╛

RouterTrieRouter
╒═══════════════════════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╕
│ Metric                    │      p0 │     p25 │     p50 │     p75 │     p90 │     p99 │    p100 │ Samples │
╞═══════════════════════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╡
│ Malloc (total) *          │      22 │      22 │      22 │      22 │      22 │      22 │      22 │     332 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Throughput (# / s) (K)    │     341 │     339 │     337 │     335 │     330 │     320 │     306 │     332 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (total CPU) (ns) *   │    2956 │    2972 │    2984 │    3000 │    3043 │    3119 │    3255 │     332 │
╘═══════════════════════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╛

RouterTrieRouterParameters
╒═══════════════════════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╕
│ Metric                    │      p0 │     p25 │     p50 │     p75 │     p90 │     p99 │    p100 │ Samples │
╞═══════════════════════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╡
│ Malloc (total) *          │      21 │      21 │      21 │      21 │      21 │      21 │      21 │     327 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Throughput (# / s) (K)    │     335 │     333 │     332 │     330 │     326 │     319 │     301 │     327 │
├───────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (total CPU) (ns) *   │    3003 │    3025 │    3035 │    3047 │    3086 │    3146 │    3302 │     327 │
╘═══════════════════════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╛
```

Reply:
The `RouterTrieRouter` has gotten worse in P90 and higher.